include/miopen/miopen.h

```diff
@@ -538,6 +538,15 @@ MIOPEN_EXPORT miopenStatus_t miopenCreateTensorDescriptor(miopenTensorDescriptor
 MIOPEN_EXPORT miopenStatus_t miopenSet4dTensorDescriptor(
     miopenTensorDescriptor_t tensorDesc, miopenDataType_t dataType, int n, int c, int h, int w);
 
+/*! @brief Set the layout to a tensor descriptor
+ *
+ * API for setting the layout to a tensor descriptor
+ * @param tensorDesc Tensor descriptor type (input)
+ * @param layout     Tensor layout (input)
+ */
+MIOPEN_EXPORT miopenStatus_t miopenSetTensorDescriptorLayout(miopenTensorDescriptor_t tensorDesc,
```

Comment:
We should not have this API. The layout can be set by using the strides or we can provide setters for the common layouts, such as `miopenSetCNHWTensorDescriptor` or `miopenSetNHWCTensorDescriptor`.

Reply:
I don't understand the underlying reason why this should not be an API. The other options are good sure, but what is the actual objection?

Reply:
The main problem is we have two ways to represent the layout. We can already specify CNHW layout with `miopenSetTensorDescriptor(tensorDesc, dataType, 4, {N, C, H, W}.begin(), {H*W, N*H*W, W, 1}.begin())`. Although this may not always be the most convenient for some users, which is why we should provide `miopenSetCNHWTensorDescriptor`. 

Using a string for the layout leads to us erroneously using the string to determine the layout instead of using the actual layout of the tensor. We could make the layout string and the actual layout of the tensor work, but this will double our test coverage. 

Also, MIGraphX will most likely specify the layout using the strides and not a layout string as it would be too cumbersome to infer.

Reply:
+1 to Paul, I do not like strings here due to many reasons.

However I do not understand how `miopenSetTensorDescriptor(tensorDesc, dataType, 4, {N, C, H, W}.begin(), {H*W, N*H*W, W, 1}.begin())` can do the job. This doesn't look like a C code to me. Perhaps `{N, C, H, W}.begin()` and `{H*W, N*H*W, W, 1}.begin())` represent beginnings of buffers, but I do not see how the library can derive information about layout from two pointers. Maybe I just need to sleep some time? ;)

Reply:
>  This doesn't look like a C code to me.

This is not C code its C++ code. The `{N, C, H, W}` creates an initializer list and `.begin()` returns a pointer to it, so it can be passed to the C API. I was just trying to shorten it to one line. Another way to write it would be(assuming `N`, `C`, `H`, and `W` are constants):

```cpp
std::vector<int> lens = {N, C, H, W};
std::vector<int> strides = {H*W, N*H*W, W, 1};
miopenSetTensorDescriptor(tensorDesc, dataType, 4, lens.data(), strides.data());
```

> but I do not see how the library can derive information about layout from two pointers.

It derives the layout from the strides. As I showed below with the `GetLayout` function, `sort_permutation` will return the permutation `0,1,2,3` for NCHW layout and in the example above for CNHW it will return the permutation `1,0,2,3`.

If we pass the labels in as a string to `GetLayout` for the default permutation(ie `0,1,2,3`) then it will put the labels in the order of its layout. So `GetLayout("NCHW")` will return `NCHW` with permutation of `0,1,2,3` and it will return `CNHW` for permutation of `1,0,2,3`.

Reply:
Thanks for the comments guys. It makes total sense after looking through especially @pfultz2 comments and suggestions. Using strides to infer the layout does make it less ambiguous and make it able to take advantage the existing API as well. Will push a new commit to address this.

Reply:
@pfultz2 Thanks! it became clear to me in the morning, indeed :facepalm: